cmd-dynamic.c

```diff
@@ -0,0 +1,473 @@
+		set_env_to_file(env_fd, "UFTRACE_DEPTH", buf);
+	}
+
+	if (opts->max_stack != OPT_RSTACK_DEFAULT) {
+		snprintf(buf, sizeof(buf), "%d", opts->max_stack);
+		set_env_to_file(env_fd, "UFTRACE_MAX_STACK", buf);
+	}
+
+	if (opts->threshold) {
+		snprintf(buf, sizeof(buf), "%"PRIu64, opts->threshold);
+		set_env_to_file(env_fd, "UFTRACE_THRESHOLD", buf);
+	}
+
+	if (opts->libcall) {
+		set_env_to_file(env_fd, "UFTRACE_PLTHOOK", "1");
+
+		if (opts->want_bind_not) {
+			/* do not update GOTPLT after resolving symbols */
+			set_env_to_file(env_fd, "LD_BIND_NOT", "1");
+		}
+
+		if (opts->nest_libcall)
+			set_env_to_file(env_fd, "UFTRACE_NEST_LIBCALL", "1");
+	}
+
+	if (strcmp(opts->dirname, UFTRACE_DIR_NAME))
+		set_env_to_file(env_fd, "UFTRACE_DIR", opts->dirname);
+
+	if (opts->bufsize != SHMEM_BUFFER_SIZE) {
+		snprintf(buf, sizeof(buf), "%lu", opts->bufsize);
+		set_env_to_file(env_fd, "UFTRACE_BUFFER", buf);
+	}
+
+	if (opts->logfile) {
+		snprintf(buf, sizeof(buf), "%d", fileno(logfp));
+		set_env_to_file(env_fd, "UFTRACE_LOGFD", buf);
+	}
+
+	snprintf(buf, sizeof(buf), "%d", pfd);
+	set_env_to_file(env_fd, "UFTRACE_PIPE", buf);
+	set_env_to_file(env_fd, "UFTRACE_SHMEM", "1");
+
+	if (debug) {
+		snprintf(buf, sizeof(buf), "%d", debug);
+		set_env_to_file(env_fd, "UFTRACE_DEBUG", buf);
+		set_env_to_file(env_fd, "UFTRACE_DEBUG_DOMAIN", build_debug_domain_string());
+	}
+
+	if (opts->disabled)
+		set_env_to_file(env_fd, "UFTRACE_DISABLED", "1");
+
+	if (log_color == COLOR_ON) {
+		snprintf(buf, sizeof(buf), "%d", log_color);
+		set_env_to_file(env_fd, "UFTRACE_COLOR", buf);
+	}
+
+	snprintf(buf, sizeof(buf), "%d", demangler);
+	set_env_to_file(env_fd, "UFTRACE_DEMANGLE", buf);
+
+	if ((opts->kernel || has_kernel_event(opts->event)) &&
+			check_kernel_pid_filter()) {
+		set_env_to_file(env_fd, "UFTRACE_KERNEL_PID_UPDATE", "1");
+	}
+
+	if (opts->script_file)
+		set_env_to_file(env_fd, "UFTRACE_SCRIPT", opts->script_file);
+
+	if (opts->lib_path)
+		snprintf(buf, sizeof(buf), "%s/libmcount/", opts->lib_path);
+	else
+		buf[0] = '\0';  /* to make strcat() work */
+
+	strcat(buf, "libmcount-dynamic.so");
+	pr_dbg("using %s library for tracing\n", buf);
+
+	old_preload = getenv("LD_PRELOAD");
+	if (old_preload) {
+		size_t len = strlen(buf) + strlen(old_preload) + 2;
+		char *preload = xmalloc(len);
+
+		snprintf(preload, len, "%s:%s", buf, old_preload);
+		set_env_to_file(env_fd, "LD_PRELOAD", preload);
+		free(preload);
+	}
+	else {
+		set_env_to_file(env_fd, "LD_PRELOAD", buf);
+	}
+
+	// The process ID to inject the shared object.
+	snprintf(buf, sizeof(buf), "%d", getpid());
+	set_env_to_file(env_fd, "UFTRACE_PID", buf);
+
+	// The below code should be placed at the end.
+	set_env_to_file(env_fd, "XRAY_OPTIONS", "patch_premain=false");
+	close(env_fd);
+}
+
+// settings for using dynamic tracing feature.
+void setup_uftrace_environ(struct opts *opts, int pfd)
+{
+	make_uftrace_environ_file(opts, pfd);
```

Comment:
I thought it is enough that if there is one of the two functions `setup_uftrace_environ` and `make_uftrace_environ_file`, no ?

Reply:
@Taeung you're right. I'm considering that remove setup_uftrace_environ.